File: android/src/ti/airship/utils/PluginStore.java
/* Copyright Urban Airship and Contributors */

package ti.airship.utils;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;

/**
 * Stores shared preferences and checks preference-dependent state.
 */
public class PluginStore {
    private static final String SHARED_PREFERENCES_FILE = "ti.airship";

    private static PluginStore sharedInstance;
    private static final Object sharedInstanceLock = new Object();

    private static final String AIRSHIP_CONFIG = "AIRSHIP_CONFIG";
    private static final String PREFERENCE_CENTER_CUSTOM_UI_PREFIX = "PREFERENCE_CENTER_CUSTOM_UI-";

    private SharedPreferences preferences;

    private PluginStore(@NonNull Context context) {
        this.preferences = context.getSharedPreferences(SHARED_PREFERENCES_FILE, Context.MODE_PRIVATE);
    }

    public static PluginStore shared(@NonNull Context context) {
        synchronized (sharedInstanceLock) {
            if (sharedInstance == null) {
                sharedInstance = new PluginStore(context);
            }
            return sharedInstance;
        }
    }

    @Nullable
    public JsonMap getAirshipConfig() {
        String config = preferences.getString(AIRSHIP_CONFIG, null);
        if (config == null) {
            return JsonMap.EMPTY_MAP;
        }

        try {
            return JsonValue.parseString(config).getMap();
        } catch (JsonException e) {
            PluginLogger.error("Failed to parse config.", e);
            return null;
        }
    }

    public void setAirshipConfig(JsonMap config) {
        preferences.edit().putString(AIRSHIP_CONFIG, config.toString()).apply();
    }

    public void setUseCustomPreferenceCenter(@NonNull String preferenceId, boolean useCustom) {
        preferences.edit()
                .putBoolean(useCustomPreferenceCenterKey(preferenceId), useCustom)
                .apply();
    }

    public boolean getUseCustomPreferenceCenter(@NonNull String preferenceId) {
        return preferences.getBoolean(useCustomPreferenceCenterKey(preferenceId), false);
    }

    private static String useCustomPreferenceCenterKey(String preferenceId) {
        return PREFERENCE_CENTER_CUSTOM_UI_PREFIX + preferenceId;
    }
}

File: android/src/ti/airship/utils/PluginStore.kt
/* Copyright Urban Airship and Contributors */

package ti.airship.utils

import android.content.Context
import android.content.SharedPreferences
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue

/**
 * Stores shared preferences and checks preference-dependent state.
 */
class PluginStore private constructor(context: Context) {
    companion object {
        private const val SHARED_PREFERENCES_FILE = "ti.airship"
        private const val AIRSHIP_CONFIG = "AIRSHIP_CONFIG"
        private const val PREFERENCE_CENTER_CUSTOM_UI_PREFIX = "PREFERENCE_CENTER_CUSTOM_UI-"

        @Volatile
        private var INSTANCE: PluginStore? = null

        fun shared(context: Context): PluginStore =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: PluginStore(context).also { INSTANCE = it }
            }


        private fun useCustomPreferenceCenterKey(preferenceId: String): String {
            return PREFERENCE_CENTER_CUSTOM_UI_PREFIX + preferenceId
        }
    }

    private val preferences: SharedPreferences

    init {
        preferences = context.getSharedPreferences(SHARED_PREFERENCES_FILE, Context.MODE_PRIVATE)
    }

    var airshipConfig: JsonMap?
        get() {
            val config = preferences.getString(AIRSHIP_CONFIG, null)
                ?: return JsonMap.EMPTY_MAP
            return try {
                JsonValue.parseString(config).map
            } catch (e: JsonException) {
                PluginLogger.error("Failed to parse config.", e)
                null
            }
        }
        set(value) {
            preferences.edit().putString(AIRSHIP_CONFIG, value.toString()).apply()
        }

    fun setUseCustomPreferenceCenter(preferenceId: String, useCustom: Boolean) {
        preferences.edit()
            .putBoolean(useCustomPreferenceCenterKey(preferenceId), useCustom)
            .apply()
    }

    fun getUseCustomPreferenceCenter(preferenceId: String): Boolean {
        return preferences.getBoolean(useCustomPreferenceCenterKey(preferenceId), false)
    }


}